[Demangling] Fix duplicate _gCRAnnotations symbol problems.

al45tair opened the pull request:
Moved the `_gCRAnnotations` declarations to their own object module, which will help to avoid duplicate symbol problems (at least with `.a` files).

Also tweaked things to make it so that the demangler and runtime versions of the message setting code will interoperate (and so that they'll interoperate better with other implementations that might creep in from somewhere, like the one in LLVMSupport).

rdar://91095592

al45tair committed:
[Demangling] Fix duplicate _gCRAnnotations symbol problems.

Moved the _gCRAnnotations declarations to their own object module,
which will help to avoid duplicate symbol problems (at least with .a
files).

Also tweaked things to make it so that the demangler and runtime
versions of the message setting code will interoperate (and so that
they'll interoperate better with other implementations that might
creep in from somewhere, like the one in LLVMSupport).

rdar://91095592

al45tair commented:
@swift-ci Please smoke test

mikeash commented on stdlib/public/runtime/Errors.cpp:
```diff
@@ -254,47 +259,34 @@ void swift::printCurrentBacktrace(unsigned framesToSkip) {
-  crashlogLock.unlock();
+  oldMessage = std::atomic_load_explicit(
+    (volatile std::atomic<char *> *)&gCRAnnotations.message,
+    std::memory_order_relaxed);
```
Not that it's going to matter much in practice, but this should be `SWIFT_MEMORY_ORDER_CONSUME`.

al45tair committed:
[Demangling] Use SWIFT_MEMORY_ORDER_CONSUME.

Also rearrange the #ifs to make the two pieces of code match more
closely.

rdar://91095592

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test platform linux

al45tair commented:
@swift-ci Please smoke test Linux platform
